### ISSUE TITLE
Rule Proposal: lines-around-commonjs-default-exports

### ISSUE THREAD

**Issue by mysticatea** [eslint/eslint#6070]
From [requirePaddingNewLinesBeforeExport](http://jscs.info/rule/requirePaddingNewLinesBeforeExport) and [disallowPaddingNewLinesBeforeExport](http://jscs.info/rule/disallowPaddingNewLinesBeforeExport).

This rule will require/disallow blank lines before/after the assignment of `module.exports`.

``` js
{
    "lines-around-commonjs-default-exports": ["error", "always" or "never"],
    // or
    "lines-around-commonjs-default-exports": ["error", {
        "after": "always" or "never",
        "before": "always" or "never"
    }]
}
```

**Valid:**

``` js
/*eslint lines-around-commonjs-default-exports: ["error", "always"]*/
var a = 2;

module.exports = a;

if (cond) {
   module.exports = a;  // ignore at the first/last of a block. this is warned by padded-blocks
}
```

``` js
/*eslint lines-around-commonjs-default-exports: ["error", "never"]*/
var a = 2;
module.exports = a;
if (cond) {

   module.exports = a;  // ignore at the first/last of a block. this is warned by padded-blocks

}
```

**Invalid:**

``` js
/*eslint lines-around-commonjs-default-exports: ["error", "always"]*/
var a = 2;
module.exports = a;
foo();
```

``` js
/*eslint lines-around-commonjs-default-exports: ["error", "never"]*/
var a = 2;

module.exports = a;

foo();
```

**Comment by ilyavolodin** [comment]
Should the be made more generic to include ES6 module exports as well?

**Comment by mysticatea** [comment]
I think we should separate rules between ES6 modules and Commonjs's.
We have rules for each `require()` and ES6 Modules in separately.

**Comment by kaicataldo** [comment]
Is there a precedent for the behavior of ignoring a specific case if another rule covers it? Seems like it could be surprising behavior to not warn when `module.exports` is either at the beginning or end of a block.

**Comment by platinumazure** [comment]
@kaicataldo Well, according to the [Core Rule Guidelines](http://eslint.org/docs/developer-guide/contributing/new-rules#core-rule-guidelines), we need to avoid conflicts between rules. Unfortunately, that means some rules have to do an awkward dance around each other. One rule needs to be a gentleman and give way to the other. That's a crappy metaphor but hopefully you get what I'm saying. :stuck_out_tongue_winking_eye: 

In this case, either the new rule can cede ground to `padded-blocks`, or vice versa. I lean towards this rule ceding to `padded-blocks` because the beginning of the block can contain anything (directives or other statements). So I would prefer that multiple rules make one exception for block start and end, because that's easy to follow, rather than one rule (`padded-blocks`) make exceptions for many kinds of first and last statements.

Just my two cents.

**Comment by kaicataldo** [comment]
@platinumazure Sure - isn't the alternative to have an option that relaxes the rule for this case, though, rather than having it ignore that case by default? Seems like it could achieve the same goal without having potentially unexpected behavior in one case. Just a thought :)

**Comment by vitorbal** [comment]
I'm also curious if a relaxing option wouldn't be better, as I think I agree with @kaicataldo here. From a user's point of view, if I enable "lines-around-commonjs-default-exports" and I don't see an error in one specific case, I would be confused.

**Comment by vitorbal** [comment]
@kaicataldo @platinumazure @mysticatea sorry to bother you, do you guys have any strong feelings against making a relaxing option for the case where the `module.exports` is the first/last statement of a block?

Was hoping to keep the ball rolling for this one as I was looking to contribute to the JSCS compatibility milestone and this rule seems to be unclaimed so far :-)

**Comment by kaicataldo** [comment]
Since my comment, I wrote the `lines-around-directive` rule in which we decided to go for not checking at the beginning or end of blocks by default (so the opposite of what I argued for earlier in this thread). Given that, for consistency, I think we should do the same thing here, but I think we should make sure to outline this behavior in the docs and make it very clear. [Here](https://github.com/eslint/eslint/commit/c1f0d76e5e41fd032fa7190577da68a4cdedde4e#diff-1629a63b4c5edadb1b20188849374555R29) is an example of what I'm thinking of.

So it sounds like @platinumazure and I are now in agreement! If you want to champion, feel free to reach out and get some thumbs from other team members. Here's mine :+1:

**Comment by mysticatea** [comment]
I'm sorry that I had lacked response for this issue.

I think good to avoid confliction with `padded-blocks` rule by default and no option.
As my memory, we have been making rules as so, except `lines-around-comment` rule.

**Comment by vitorbal** [comment]
Thanks @kaicataldo and @mysticatea for the quick replies!

In that case, I would like to champion this rule (and I'm also willing to implement it). Counting @kaicataldo 's thumbs up we need two more now. @mysticatea @platinumazure any of you gentlemen interested? 😃

**Comment by mysticatea** [comment]
Honestly, I'm not sure why `module.exports` is special.

- What about `exports.foo`?
- Actually, is it enough if `lines-around-block`-like rule (which covers function expressions) exists?

**Comment by vitorbal** [comment]
@mysticatea those are good points.
- I think we should address `exports.foo` in this rule as well.
- do you mean expression statements or really function expressions? could you show an example of how you would imagine this rule?

**Comment by mysticatea** [comment]
> do you mean expression statements or really function expressions? could you show an example of how you would imagine this rule?

In my head, I wonder what is the difference between "assignments to `module.exports`" and "assignments to local variables".

```js
module.exports = () => {
}

foo = () => {
}

exports.a = 1
exports.b = 2
exports.c = 3

a = 1
b = 2
c = 3
```

Maybe, if someone wants blank lines around `module.exports = () => {\n}`, I guess they wants blank lines around `foo = () => {\n}`. So my concern is that this rule might be almost overlapping to `lines-around-block`-like rule.

Of course, there is a difference since this rule is focusing on `module.exports`, but I'm wondering why this rule is focusing on `module.exports`.

Note: I don't object this rule since this needs for JSCS Compatibility. Just I have a question.

**Comment by vitorbal** [comment]
I think this rule focuses on `module.exports` purely because some people might opt to have blank lines around things that are exported from a module, but not for every expression statement, for purely stylistic reasons.

I think the biggest value of us doing this rule would be because of the JSCS compatibility goal, otherwise I would argue that we should have a more generalized rule, at least in the sense of also including ES6 exports.

I don't think I am experienced enough to tell if we should try to generalize this to every expression statement or not. But what I think is that if a JSCS user wants to migrate from the `requirePaddingNewLinesBeforeExport` to an ESLint rule, if the ESLint rule also enforces padding of expression statements in general, that could be quite a surprise. That would mean that, e.g., the following snippet would originally pass and after migration it would potentially show an error:

```js
const foo = () => {};
const bar = 2;

module.exports = foo;
```

I'd be interested to hear your opinion :) (and from the rest of the team!)

**Comment by mysticatea** [comment]
Thank you for the explanation. I got it.

If the rule checks `exports.foo` also, which is this rule's behavior?

A. Check around the sequences of exporting.

```js
// Check
exports.foo = 1
exports.bar = 2
// Check
var foo = 0
var bar = 0
// Check
exports.baz = 3
// Check
```

B. Check around each exporting.

```js
// Check
exports.foo = 1
// Check
exports.bar = 2
// Check
var foo = 0
var bar = 0
// Check
exports.baz = 3
// Check
```

**Comment by vitorbal** [comment]
Good question. I would vote for B, but I am not sure. @eslint/eslint-team any thoughts on @mysticatea 's question above?

**Comment by kaicataldo** [comment]
Maybe someone from the JSCS team can comment on the behavior of the original rule so that we can ensure compatibility. The docs only have examples of 1 `module.export`, so it's not entirely clear to me.

Either way, could it be worth making an option that allows the user to pick from the options above?

**Comment by vitorbal** [comment]
As far as I understood from the rule source, the original JSCS rules only looked for `module.exports`  and didn't care about `exports.foo = 1` format at all.
Maybe checking `exports.foo` format should be as a whole an option?

**Comment by mysticatea** [comment]
If it handles `exports`, I remove `-default` from the rule name and imagine options:

```js
{
    "lines-around-commonjs-exports": ["error", "always" or "never"],
    // or
    "lines-around-commonjs-exports": ["error", {
        "module.exports": "always" or "never" or "ignore",
        "exports": "always" or "never" or "ignore"
    }]
    // or
    "lines-around-commonjs-exports": ["error", {
        "module.exports": {
            "after": "always" or "never" or "ignore",
            "before": "always" or "never" or "ignore"
        },
        "exports": {
            "after": "always" or "never" or "ignore",
            "before": "always" or "never" or "ignore"
        }
    }]
}
```

```js
// Schema
{
    "definitions": {
        "optionValue": {
            "anyOf": [
                {"enum": ["always", "never", "ignore"]},
                {
                    "type": "object",
                    "properties": {
                        "after": {"enum": ["always", "never", "ignore"]},
                        "before": {"enum": ["always", "never", "ignore"]}
                    },
                    "additionalProperties": false
                }
            ]
        }
    },
    "anyOf": [
        {"enum": ["always", "never"]},
        {
            "type": "object",
            "properties": {
                "module.exports": {"$ref": "#/definitions/optionValue"},
                "exports": {"$ref": "#/definitions/optionValue"}
            },
            "additionalProperties": false
        }
    ]
}
```

```js
// Examples
{
    "lines-around-commonjs-exports": "error",

    "lines-around-commonjs-exports": ["error", "never"],

    "lines-around-commonjs-exports": ["error", {
        "module.exports": {"before": "always"},
        "exports": "ignore"
    }],
}
```

**Comment by TheSavior** [comment]
I wrote the original JSCS rule and my intention was to ensure a blank space before the first exported thing in the file. 

My company happens to only use `module.exports` currently, so I had no opinion on supporting exports. We only ever specify `module.exports` once in our files, which is why the tests in the JSCS rule only ever show a single assignment to `module.exports`

**Comment by kaicataldo** [comment]
I support this with the configuration suggested by @mysticatea in this comment: https://github.com/eslint/eslint/issues/6070#issuecomment-247565481

**Comment by vitorbal** [comment]
If @mysticatea and @platinumazure can get behind the configuration proposed in [the comment you mention](https://github.com/eslint/eslint/issues/6070#issuecomment-247565481), then we have enough to accept this. I'm willing to champion and implement!

**Comment by vitorbal** [comment]
Quick update: I'm planning to start work on this next week. Should have something to show by the following weekend!

**Comment by TheSavior** [comment]
@vitorbal I just want to confirm that you are still planning on working on this.

**Comment by vitorbal** [comment]
@TheSavior I have some others things I need to wrap up currently before I can start anything else, so if you're interested please feel free to take over this one!

**Comment by TheSavior** [comment]
I don't have time to do this completely, but I figured starting is better than nothing. I have started a branch with the files for this rule. I have put in the schema from @mysticatea's comment and added some tests to the test file. The test file needs a lot more work, please help contribute to it.

I've added @vitorbal and @kaicataldo as contributors. I couldn't figure out how to make it open to everyone from ESLint. It may require opening a PR for that. I'm happy to do that if you guys are willing to have a WIP PR for this.

**Comment by sstern6** [comment]
Would like to collaborate and help get this out with however is interested in working on this if thats ok?

**Comment by kaicataldo** [comment]
@sstern6 Since it doesn't sound like @vitorbal or @TheSavior have the bandwidth right now, I think you can just go for it!

**Comment by sstern6** [comment]
thanks @kaicataldo will use the skeleton @TheSavior has already created.

**Comment by vitorbal** [comment]
Go for it, @sstern6!

**Comment by TheSavior** [comment]
Thank you @sstern6 for taking this on. If you'd like to use the base I started with, I've added you as a contributor to my fork. Feel free to pull my branch to your fork or work directly there. I want to make it as easy as possible for you to see this through.

**Comment by sstern6** [comment]
in the options for this rule does it need to be "fixable"? or can we get this up without that option in the meantime @kaicataldo

**Comment by TheSavior** [comment]
imo, it's fine to ship without it being fixable. Adding that in the future should be a reasonable thing to do.

**Comment by vitorbal** [comment]
I agree that its acceptable to ship this without being fixable first

**Comment by kaicataldo** [comment]
Agreed. Rule does not have to be fixable to land and can be added in a follow-up PR.

**Comment by vitorbal** [comment]
Should we just close this in favor of #8099? /cc @mysticatea

**Comment by kaicataldo** [comment]
Oh interesting - hadn't considered how this interacts with the newline-between-statements proposal
